GowerMetric.py

```diff
@@ -0,0 +1,361 @@
+    def __init__(
+        self,
+        dtypes: np.array,
+        ratio_scale_normalization: str = "range",
+        weights: Optional[Union[list, str]] = None,
+        precomputed_weights_file: Optional[str] = None,
+    ):
+        assert weights is None or weights == 'precomputed' or weights == 'cpcc' or type(weights) == np.ndarray or type(weights) == list
+
+        self.dtypes = dtypes  # initialize with np.array of column data types
+        self.weights = weights
+        self.precomputed_weights_file = precomputed_weights_file
+        self.ranges_: List  # values of ranges in .ratio_scale() (iqr or traditional range)
+        self.h_: List  # h values in .ratio_scale()
+        self.n_features_in_: int
+        self.ratio_scale_normalization: str = ratio_scale_normalization
+
+        # Bit masks for certain column types
+        self.ratio_scale_idx = self.dtypes == DataType.RATIO_SCALE
+        self.cat_nom_idx = (self.dtypes == DataType.CATEGORICAL_NOMINAL) | (self.dtypes == DataType.BINARY_SYMMETRIC)
+        self.bin_asym_idx = self.dtypes == DataType.BINARY_ASYMMETRIC
+
+        # Sums of columns of given type
+        self.cat_nom_num = sum(self.cat_nom_idx)
+        self.ratio_scale_num = sum(self.ratio_scale_idx)
+        self.bin_asym_num = sum(self.bin_asym_idx)
+
+    def fit(self, X):
+        assert X.shape[1] == len(self.dtypes)
+
+        self.ranges_ = np.empty(X.shape[1])
+        self.h_ = np.empty(X.shape[1])
+        self.n_features_in_ = X.shape[1]
+
+        ratio_cols = X[:, self.ratio_scale_idx]
+        if self.ratio_scale_normalization == 'range':
+            self.ranges_[self.ratio_scale_idx] = np.ptp(ratio_cols, axis=0)
+        elif self.ratio_scale_normalization == 'iqr' or self.ratio_scale_normalization == 'kde':
+            q1, q3 = np.percentile(ratio_cols, [25, 75], axis=0)
+            self.ranges_[self.ratio_scale_idx] = q3 - q1
+
+            zero_ranges_idx = (self.ranges_ == 0) & (self.ratio_scale_idx)
+            self.ranges_[zero_ranges_idx] = np.ptp(X[:, zero_ranges_idx], axis=0)
+
+        if self.ratio_scale_normalization == 'kde':
+            n = X.shape[0]
+            c = 1.06
+
+            s = np.std(ratio_cols, axis=0)
+            self.h_[self.ratio_scale_idx] = (
+                c / n ** (1 / 5) * np.min([s, self.ranges_[self.ratio_scale_idx] / 1.34], axis=0)
+            )
+
+        loader = GowerMetricWeights(self)
+        if self.weights == "precomputed":
+            loader.load_weights(self.precomputed_weights_file)
+        elif self.weights == "cpcc":
+            self.weights = np.ones(self.n_features_in_)
+            loader.select_weights(X)
+
+        loader.select_number_of_clusters(X)
+
+    def __call__(
+        self,
+        vector_1: np.ndarray,
+        vector_2: np.ndarray,
+    ):
+        # TODO - no check for fit
+        # sklearn.utils.validation.check_is_fitted(self)
+        assert self.n_features_in_ == len(vector_1) | len(vector_2)
```

Comment:
Maybe do 2 asserts, if any is not passed the error message would be more readable to user